table/tables/tables.go

```diff
@@ -485,8 +486,31 @@ func FindPrimaryIndex(tblInfo *model.TableInfo) *model.IndexInfo {
 	return pkIdx
 }
 
+// CommonAddRecordCtx is used in AddRecordWithCtx to avoid memory malloc for some temp slices.
+// This is useful in lightning parse row data to key-values pairs. This can gain upto 5%  performance
+// improvement in lightning's local mode.
+type CommonAddRecordCtx struct {
+	colIDs []int64
+	row    []types.Datum
+	buffer *kv.BufferStore
+}
+
+// NewCommonAddRecordCtx create a context used for `AddRecordWithCtx`
+func NewCommonAddRecordCtx(size int, buffer *kv.BufferStore) *CommonAddRecordCtx {
+	return &CommonAddRecordCtx{
+		colIDs: make([]int64, 0, size),
+		row:    make([]types.Datum, 0, size),
+		buffer: buffer,
+	}
+}
+
 // AddRecord implements table.Table AddRecord interface.
 func (t *TableCommon) AddRecord(ctx sessionctx.Context, r []types.Datum, opts ...table.AddRecordOption) (recordID kv.Handle, err error) {
+	return t.AddRecordWithCtx(ctx, r, nil, opts...)
+}
+
+// AddRecordWithCtx implements table.Table AddRecordWithCtx interface.
+func (t *TableCommon) AddRecordWithCtx(ctx sessionctx.Context, r []types.Datum, addCtx interface{}, opts ...table.AddRecordOption) (recordID kv.Handle, err error) {
```

Comment:
I think we can put the `addCtx` into the `ctx`, then we don't need to add a method to the `Table` interface.